File: packages/virage-core/src/quality/metrics/dense-input.ts
/**
 * Component 3 — Dense Input Preparation metrics (2 metrics)
 *
 * Text Purity:        fraction of printable/non-control characters in denseText.
 * Enrichment Quality: cosine similarity between raw chunk text and enriched denseText.
 */

import type { MetricResult } from "../interfaces.js";
import { normalizeMonotonicUp01 } from "../scoring.js";

export interface DenseInputMetricsInput {
  chunks: Array<{
    denseText: string;
    sparseText?: string;
  }>;
  embedFn: (text: string) => Promise<number[]>;
}

function isPrintable(ch: string): boolean {
  const cp = ch.codePointAt(0) ?? 0;
  return cp >= 0x20 && cp !== 0x7f;
}

function computeTextPurity(denseText: string): number {
  if (denseText.length === 0) return 0;
  let printable = 0;
  for (const ch of denseText) {
    if (isPrintable(ch)) printable++;
  }
  return printable / denseText.length;
}

function cosineSim(a: number[], b: number[]): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  for (let i = 0; i < a.length; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  const denom = Math.sqrt(normA) * Math.sqrt(normB);
  return denom === 0 ? 0 : dot / denom;
}

export async function computeDenseInputMetrics(
  input: DenseInputMetricsInput,
  sampleSize = 50,
  weightOverrides: Partial<Record<string, number>> = {},
): Promise<MetricResult[]> {
  const { chunks, embedFn } = input;
  const sample = chunks.slice(0, sampleSize);

  // Text Purity: mean across sample
  const purities = sample.map((c) => computeTextPurity(c.denseText));
  const purity =
    purities.length === 0
      ? 0
      : purities.reduce((s, v) => s + v, 0) / purities.length;

  // Enrichment Quality: only meaningful when sparseText (raw) != denseText (enriched)
  const enrichedPairs = sample.filter(
    (c) => c.sparseText != null && c.sparseText !== c.denseText,
  );

  let enrichmentQuality: number | null = null;
  if (enrichedPairs.length > 0) {
    const sims: number[] = [];
    for (const chunk of enrichedPairs) {
      const [rawEmbed, enrichedEmbed] = await Promise.all([
        embedFn(chunk.sparseText!),
        embedFn(chunk.denseText),
      ]);
      sims.push(cosineSim(rawEmbed, enrichedEmbed));
    }
    enrichmentQuality = sims.reduce((s, v) => s + v, 0) / sims.length;
  }

  return [
    {
      name: "TextPurity",
      rawValue: purity,
      normalizedValue: normalizeMonotonicUp01(purity),
      weight: weightOverrides["textPurity"] ?? 1.0,
      skipped: false,
    },
    {
      name: "EnrichmentQuality",
      rawValue: enrichmentQuality ?? 0,
      normalizedValue:
        enrichmentQuality != null
          ? normalizeMonotonicUp01((enrichmentQuality + 1) / 2)
          : 0,
      weight: weightOverrides["enrichmentQuality"] ?? 1.0,
      skipped: enrichmentQuality == null,
      skipReason:
        enrichmentQuality == null
          ? "sparseText equals denseText in all sampled chunks (no enrichment detected)"
          : undefined,
    },
  ];
}

File: crates/virage-engine/src/quality/metrics/dense_input.rs
//! Component 3 — Dense Input Preparation metrics (2 metrics).
//!
//! Text Purity:        fraction of printable/non-control characters in denseText.
//! Enrichment Quality: cosine similarity between raw chunk text and enriched denseText.
//!
//! Ported from `dist/quality/metrics/dense-input.js` — IR-038.

use std::sync::Mutex;

use crate::embedders::Embedder;
use crate::quality::scoring::normalize_monotonic_up01;
use crate::quality::{MetricResult, QualityChunk};

use super::cosine_sim;

fn is_printable(ch: char) -> bool {
    let cp = ch as u32;
    cp >= 0x20 && cp != 0x7f
}

fn compute_text_purity(dense_text: &str) -> f64 {
    if dense_text.is_empty() {
        return 0.0;
    }
    let total = dense_text.chars().count();
    let printable = dense_text.chars().filter(|c| is_printable(*c)).count();
    printable as f64 / total as f64
}

pub fn compute_dense_input_metrics(
    chunks: &[QualityChunk],
    embedder: &Mutex<dyn Embedder + Send>,
    sample_size: usize,
) -> anyhow::Result<Vec<MetricResult>> {
    let sample: Vec<&QualityChunk> = chunks.iter().take(sample_size).collect();

    let purities: Vec<f64> = sample
        .iter()
        .map(|c| compute_text_purity(&c.dense_text))
        .collect();
    let purity = if purities.is_empty() {
        0.0
    } else {
        purities.iter().sum::<f64>() / purities.len() as f64
    };

    let enriched_pairs: Vec<&QualityChunk> = sample
        .iter()
        .filter(|c| !c.sparse_text.is_empty() && c.sparse_text != c.dense_text)
        .copied()
        .collect();

    let enrichment_quality = if enriched_pairs.is_empty() {
        None
    } else {
        let mut guard = embedder
            .lock()
            .map_err(|_| anyhow::anyhow!("embedder lock poisoned"))?;
        let dims = guard.dimensions();
        let mut sims = Vec::with_capacity(enriched_pairs.len());
        for chunk in &enriched_pairs {
            let texts = [chunk.sparse_text.as_str(), chunk.dense_text.as_str()];
            let flat = guard
                .embed_batch(&texts)
                .map_err(|e| anyhow::anyhow!("embed error: {e}"))?;
            let raw = &flat[0..dims];
            let enriched = &flat[dims..dims * 2];
            sims.push(cosine_sim(raw, enriched));
        }
        Some(sims.iter().sum::<f64>() / sims.len() as f64)
    };

    let results = vec![
        MetricResult::new("TextPurity", purity, normalize_monotonic_up01(purity), 1.0),
        match enrichment_quality {
            Some(v) => MetricResult::new(
                "EnrichmentQuality",
                v,
                normalize_monotonic_up01((v + 1.0) / 2.0),
                1.0,
            ),
            None => MetricResult::skipped(
                "EnrichmentQuality",
                1.0,
                "sparseText equals denseText in all sampled chunks (no enrichment detected)",
            ),
        },
    ];
    Ok(results)
}
